baseapp/abci.go

```diff
@@ -182,13 +185,22 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
 		WithHeaderHash(req.Hash).
 		WithConsensusParams(app.GetConsensusParams(app.deliverState.ctx))
 
-	// we also set block gas meter to checkState in case the application needs to
-	// verify gas consumption during (Re)CheckTx
+	// we
 	if app.checkState != nil {
 		app.checkState.ctx = app.checkState.ctx.
 			WithBlockGasMeter(gasMeter).
 			WithHeaderHash(req.Hash)
 	}
+	if app.prepareProposalState != nil {
+		app.prepareProposalState.ctx = app.prepareProposalState.ctx.
+			WithBlockGasMeter(gasMeter).
+			WithHeaderHash(req.Hash)
+	}
+	if app.processProposalState != nil {
+		app.processProposalState.ctx = app.processProposalState.ctx.
+			WithBlockGasMeter(gasMeter).
+			WithHeaderHash(req.Hash)
+	}
```

Comment:
I'm pretty sure `PrepareProposal` does not need its own state. We didn't mention it in the ADR at least.

The idea behind `PrepareProposal` is it selects N txs and then essentially runs `CheckTx` on each tx. For each tx:

1. Run the `CheckTx` flow (note, this should already use a branched state that we DO NOT commit)
1a. If tx fails, remove it from the mempool and the proposal response
1b. If tx passes, keep it in the proposal response

Note, this brings up an interesting point I haven't thought of before. If (1a) ever happens, that means we'll have additional capacity to select more txs possibly. But since we don't remove txs when we select, how do we select more?

Reply:
This was resolved by moving the `Select` API to return an iterator.